### DISCUSSION TITLE
how to use icu::collator for gb18030 encoding?

### DISCUSSION THREAD

**Discussion by qwm-db** [unicode-org/icu4x#4039] [Q&A]
as rust only support utf-8 encoding string,  and the collator api use str slice as parameter, so seems it's not feasible to pass gb18030 encoding bytes?

**Comment by sffc** [comment]
ICU4X supports Han text in UTF-8 and UTF-16. The following Han orderings are available:

- big5han
- gb2312han
- pinyin
- stroke
- unihan
- zhuyin

Note that big5han and gb2312han need to be explicitly enabled in datagen. When used, they collate the text in the order that would result if the text were in either of those two encodings. However, they are not recommended for modern use; one of the other orderings, such as "pinyin" or "stroke", should produce the best quality collation results.

**Reply by qwm-db** [reply]
thanks @sffc  for shedding light here.

sorry, two follow up questions, many thanks in advance!
1) double confirm here: so the input string is in either utf-8 or utf-16 encoding? not in gb18030 encoding?
2) if I wana to use icu_collator crate, where can I get gb18030 data to setup data_provider? 

pub fn [try_new_unstable](https://docs.rs/icu_collator/latest/icu_collator/struct.Collator.html#method.try_new_unstable)<D>(
    data_provider: [&D](https://doc.rust-lang.org/nightly/std/primitive.reference.html),
    locale: &[DataLocale](https://docs.rs/icu_provider/1.2.0/x86_64-unknown-linux-gnu/icu_provider/request/struct.DataLocale.html),
    options: [CollatorOptions](https://docs.rs/icu_collator/latest/icu_collator/struct.CollatorOptions.html)
) -> [Result](https://doc.rust-lang.org/nightly/core/result/enum.Result.html)<Self, [CollatorError](https://docs.rs/icu_collator/latest/icu_collator/enum.CollatorError.html)>

**Reply by markusicu** [reply]
Correct -- modern i18n libraries necessarily operate on Unicode text.
When you have text in a legacy codepage, you need to convert to/from Unicode.
You can use ICU4C/ICU4J or likely one of several Rust libraries to convert from GB18030 to Unicode.

**Reply by sffc** [reply]
You can use [encoding_rs](https://docs.rs/encoding_rs/latest/encoding_rs/), built by ICU4X contributor @hsivonen, to convert between GB-18030 and UTF-8.

**Reply by qwm-db** [reply]
again many thanks @markusicu and @sffc for your kind help.

btw where can I find the data of gb18030 datakey to setup the DataProvider?
https://docs.rs/icu_provider/1.2.0/icu_provider/trait.DataProvider.html

**Reply by hsivonen** [reply]
There is no gb18030 key for `DataProvider`. The encoding_rs crate contains the mapping tables from gb18030 to UTF-8 independently of ICU4X data. `pinyin` is the default collation order when requesting `zh` as the locale.
